dfp/decimal/decimal

```diff
@@ -180,6 +180,286 @@ public:
+      if (!__sav)
+	{
+	  __err = ios_base::failbit;
+	  return;
+	}
+
+      char* __sanity;
+      __v = strtod128(__s, &__sanity);
+
+      // _GLIBCXX_RESOLVE_LIB_DEFECTS
+      // 23. Num_get overflow result.
+      if (__sanity == __s || *__sanity != '\0')
+	{
+	  __v = decimal::decimal128 {0.0dl};
+	  __err = ios_base::failbit;
+	}
+       else if (
+	  __v == decimal::decimal128 {__builtin_infd128()}
+	  || __v == decimal::decimal128 {-__builtin_infd128()} )
+	{
+	  if (__v > decimal::decimal128 {0.0dl})
+	    __v = DEC128_MAX;
+	  else
+	    __v = -DEC128_MAX;
+	  __err = ios_base::failbit;
+	}
+
+      setlocale(LC_ALL, __sav);
+      delete [] __sav;
+    }
+
+
+
+  template<typename _CharT, typename _InIter = std::istreambuf_iterator<_CharT>>
+  class dfp_num_get : public num_get<_CharT, _InIter> {
+      typedef _InIter iter_type;
+
+  protected:
+      virtual iter_type
+      do_get(iter_type, iter_type, ios_base&, ios_base::iostate&,
+        decimal::decimal32&) const;
+
+      virtual iter_type
+      do_get(iter_type, iter_type, ios_base&, ios_base::iostate&,
+        decimal::decimal64&) const;
+
+      virtual iter_type
+      do_get(iter_type, iter_type, ios_base&, ios_base::iostate&,
+        decimal::decimal128&) const;
+
+  public:
+      iter_type
+      get(iter_type in, iter_type end, ios_base& io,
+     ios_base::iostate& err, decimal::decimal32& v) const
+      { return this->do_get(in, end, io, err, v); }
+
+      iter_type
+      get(iter_type in, iter_type end, ios_base& io,
+          ios_base::iostate& err, decimal::decimal64& v) const
+      { return this->do_get(in, end, io, err, v); }
+
+      iter_type
+      get(iter_type in, iter_type end, ios_base& io,
+          ios_base::iostate& err, decimal::decimal128& v) const
+      { return this->do_get(in, end, io, err, v); }
+  };
+
+  template<typename _CharT, typename _InIter>
+    _InIter
+    dfp_num_get<_CharT, _InIter>::
+    do_get(iter_type __beg, iter_type __end, ios_base& __io,
+           ios_base::iostate& __err, decimal::decimal32& __v) const
+    {
+      string __xtrc;
+      __xtrc.reserve(32);
+      __beg = this->_M_extract_float(__beg, __end, __io, __err, __xtrc);
+      std::__convert_to_v(__xtrc.c_str(), __v, __err, locale::facet::_S_get_c_locale());
+      if (__beg == __end)
+        __err |= ios_base::eofbit;
+      return __beg;
+    }
+
+  template<typename _CharT, typename _InIter>
+    _InIter
+    dfp_num_get<_CharT, _InIter>::
+    do_get(iter_type __beg, iter_type __end, ios_base& __io,
+           ios_base::iostate& __err, decimal::decimal64& __v) const
+    {
+      string __xtrc;
+      __xtrc.reserve(32);
+      __beg = this->_M_extract_float(__beg, __end, __io, __err, __xtrc);
+      std::__convert_to_v(__xtrc.c_str(), __v, __err, locale::facet::_S_get_c_locale());
+      if (__beg == __end)
+        __err |= ios_base::eofbit;
+      return __beg;
+    }
+
+  template<typename _CharT, typename _InIter>
```

Comment:
Is it possible to parameterize the decimal type here to avoid duplicating this for each decimal type?

Reply:
These are virtual functions, which cannot be templates.